staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+// getCRDSchemaForVersion returns the validation schema for given version in given CRD.
+func getCRDSchemaForVersion(crd *apiextensions.CustomResourceDefinition, version string) (*apiextensions.CustomResourceValidation, error) {
+	for _, v := range crd.Spec.Versions {
+		if version != v.Name {
+			continue
+		}
+		if crd.Spec.Validation != nil && v.Schema != nil {
+			return nil, fmt.Errorf("malformed CustomResourceDefinition %s version %s: top-level and per-version schemas must be mutual exclusive", crd.Name, version)
+		}
+		if v.Schema != nil {
+			return v.Schema, nil
+		}
+		return crd.Spec.Validation, nil
```

Comment:
are we doing override here? we agreed not to do overrides. That means in getCRDSchemaForVersion we should call the other helper to see if there is any per-version Schema. If there is we should return `v.Schema` even if it is `nil`.

Reply:
No, it's not doing override. I think the auto-clear strategy makes sure that both ways work the same. In the case that you described (`v.Schema` is `nil` while there exists other per-version schema), `crd.Spec.Validation` will also be `nil`, so `getCRDSchemaForVersion` returns `nil` in the end.

Reply:
You are right. Still to anybody reading this code, it look like we are overriding. It would be a cleaner code if we apply the logic here too and it wouldn't hurt.

Reply:
or you can check if top level field is not `nil` and return it at the start of the method with a comment explaining why if it is not nil, it means all per versions will be. I still prefer the first approach though :)